src/Microsoft.Framework.PackageManager/Algorithm/DFPreOrderWalker.cs

```diff
@@ -0,0 +1,99 @@
+        private Func<T, IEnumerable<T>> _getChildren;
+
+        public DFPreOrderWalker(Func<WalkContext, IEnumerable<WalkContext>, bool> visit,
+                                Func<T, IEnumerable<T>> getChildren)
+        {
+            _visit = visit;
+            _getChildren = getChildren;
+        }
+
+        public void Walk(T node)
+        {
+            Walk(new WalkContext(node, true), new Stack<WalkContext>());
+        }
+
+        private void Walk(WalkContext current, Stack<WalkContext> previous)
+        {
+            if (_visit(current, previous))
+            {
+                previous.Push(current);
+
+                var childrenInArray = _getChildren(current.Node)?.ToArray() ?? new T[] { };
+                for (int i = 0; i < childrenInArray.Length; ++i)
+                {
+                    Walk(new WalkContext(childrenInArray[i], i == childrenInArray.Length - 1), previous);
+                }
+
+                previous.Pop();
+            }
+        }
+
+        public class WalkContext
+        {
+            internal WalkContext(T node, bool lastInSiblings)
+            {
+                Node = node;
+                LastInSiblings = lastInSiblings;
+            }
+
+            public T Node { get; }
+
+            public bool LastInSiblings { get; }
+        }
+    }
+
+    /// <summary>
+    /// A specialized version of DFPreOrderWalker.
+    /// 
+    /// It is simplified for depth-first traversal of IGraphNode base graph. The 
+    /// detail like WalkContext and IVertext are largely hidden from consumer, 
+    /// except the Walk function still accepts root of the graph as type IGraphNode.
+    /// </summary>
+    /// <typeparam name="T">Type of the grahp node item</typeparam>
+    public class DFPreOrderVertexWalker<T>
+    {
+        private DFPreOrderWalker<IGraphNode<T>> _algo;
+
+        /// <summary>
+        /// Constructs a DFPreOderVertexWalker
+        /// </summary>
+        /// <param name="visit">
+        /// The function to visit graph node.
+        /// 
+        /// First parameter of type T is the item currently being visited. 
+        /// Second parameter of type boolean represents whether it is the 
+        ///   last node among its siblings. 
+        /// The third parameter of a list of Tuple represent the item and 
+        ///   whether each is the last node among its siblings from the 
+        ///   path from current node to the root node. 
+        /// If the function returns false the walking will be terminated 
+        ///   on this path.
+        /// </param>
+        public DFPreOrderVertexWalker(Func<T, bool, IEnumerable<Tuple<T, bool>>, bool> visit)
+        {
+            _algo = new DFPreOrderWalker<IGraphNode<T>>(
+                (current, previouses) =>
+                {
+                    return visit(current.Node.Item, current.LastInSiblings, previouses.Select(c => new Tuple<T, bool>(c.Node.Item, c.LastInSiblings)).ToArray());
```

Comment:
Maybe assign the `previouses.Select(c...` to a separate variable for readability.